### PR TITLE
feat: Rows dependencies

### PR DESCRIPTION
**Pull request by jajoe**
Add a function to find the dependencies between rows.

### CODE REVIEW

File: src/linearDependencies.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        }
+        return returnArray.to1DArray();
+    }
+}
+
+/**
+ * Creates a matrix which represents the dependencies between rows. If a row is a linear combinaison of others rows, the result will be a row with the coefficients of this combinaison. So we will have as result a matrix, with for each row its linear combinaison (or a row of 0 if this row is independent.
+ * For example : for A = [[2, 0, 0, 1], [0, 1, 6, 0], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]], the result will be [ 0, 0, 0, 0, 0 ], [ 0, 0, 0, 4, 1 ], [ 0, 0, 0, 0, 0 ],[ 0, 0.25, 0, 0, -0.25 ], [ 0, 1, 0, -4, 0 ], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]])
+ * @param {Matrix} the matrix
```
Comment (maasencioh) [inline src/linearDependencies.js]: `@param {Matrix} A - the matrix`

File: src/linearDependencies.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import Matrix from './matrix';
+import SingularValueDecomposition from './dc/svd';
+import max from 'ml-array-max';
+
+// function used by rowsDependencies
+function xrange(n, exception) {
```
Comment (maasencioh) [inline src/linearDependencies.js]: https://www.npmjs.com/package/lodash.range

File: src/linearDependencies.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Creates a matrix which represents the dependencies between rows. If a row is a linear combinaison of others rows, the result will be a row with the coefficients of this combinaison. So we will have as result a matrix, with for each row its linear combinaison (or a row of 0 if this row is independent.
+ * For example : for A = [[2, 0, 0, 1], [0, 1, 6, 0], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]], the result will be [ 0, 0, 0, 0, 0 ], [ 0, 0, 0, 4, 1 ], [ 0, 0, 0, 0, 0 ],[ 0, 0.25, 0, 0, -0.25 ], [ 0, 1, 0, -4, 0 ], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]])
+ * @param {Matrix} the matrix
+ * @param {number} thresholdValue - if an absolute value is inferior to this threshold, it will equals zero
+ * @param {number} thresholdError - if the error is inferior to that threshold, we say that the linear combinaison found is accepted, so the row is dependent from other rows
+ * @param {Matrix} the matrix which represents the dependencies between rows.
+ */
+
+export function rowsDependencies(A, thresholdValue = 10e-10, thresholdError = 10e-10) {
```
Comment (maasencioh) [inline src/linearDependencies.js]: options should be an single object

File: src/linearDependencies.js [hunk 4]
```diff
@@ -31,13 +31,18 @@ function dependenciesOneRow(error, matrix, index, thresholdValue = 10e-10, thres
 /**
  * Creates a matrix which represents the dependencies between rows. If a row is a linear combinaison of others rows, the result will be a row with the coefficients of this combinaison. So we will have as result a matrix, with for each row its linear combinaison (or a row of 0 if this row is independent.
  * For example : for A = [[2, 0, 0, 1], [0, 1, 6, 0], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]], the result will be [ 0, 0, 0, 0, 0 ], [ 0, 0, 0, 4, 1 ], [ 0, 0, 0, 0, 0 ],[ 0, 0.25, 0, 0, -0.25 ], [ 0, 1, 0, -4, 0 ], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]])
- * @param {Matrix} the matrix
- * @param {number} thresholdValue - if an absolute value is inferior to this threshold, it will equals zero
- * @param {number} thresholdError - if the error is inferior to that threshold, we say that the linear combinaison found is accepted, so the row is dependent from other rows
+ * @param {Matrix} Matrix - the matrix
```
Comment (maasencioh) [inline src/linearDependencies.js]: the name of the parameter is not `Matrix` in your code, it's `A`

File: src/linearDependencies.js [hunk 5]
```diff
@@ -31,13 +31,18 @@ function dependenciesOneRow(error, matrix, index, thresholdValue = 10e-10, thres
...
+ * @param {Object} options - includes thresholdValue and thresholdError.
```
Comment (maasencioh) [inline src/linearDependencies.js]:
```
@param {Object} [options] - description
@param {number} [options.thresholdValue] - description
@param {number} [options.thresholdError] - description
```

**Comment by stropitek** [comment]
Can you add test cases?

File: src/linearDependencies.js (hunk 2 above)
Comment (targos) [inline src/linearDependencies.js]: lodash.range doesn't have the exception option

**Comment by targos** [comment]
package-lock.json shouldn't be removed

File: src/index.js [hunk 6]
```diff
@@ -6,6 +6,7 @@ export {default as WrapperMatrix2D} from './wrap/WrapperMatrix2D';
 export {default as WrapperMatrix1D} from './wrap/WrapperMatrix1D';
 
 export {solve, inverse} from './decompositions';
+export {rowsDependencies} from './linearDependencies';
```
Comment (targos) [inline src/index.js]: I think linearDependencies looks better for a name 🤷‍♂️

File: src/linearDependencies.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+/**
+ * Creates a matrix which represents the dependencies between rows. If a row is a linear combinaison of others rows, the result will be a row with the coefficients of this combinaison. So we will have as result a matrix, with for each row its linear combinaison (or a row of 0 if this row is independent.
+ * For example : for A = [[2, 0, 0, 1], [0, 1, 6, 0], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]], the result will be [ 0, 0, 0, 0, 0 ], [ 0, 0, 0, 4, 1 ], [ 0, 0, 0, 0, 0 ],[ 0, 0.25, 0, 0, -0.25 ], [ 0, 1, 0, -4, 0 ], [0, 3, 0, 1], [0, 0, 1, 0], [0, 1, 2, 0]])
+ * @param {Matrix} A - the matrix
+ * @param {Object} [options] includes thresholdValue and thresholdError.
+ * @param {number} [options.thresholdValue] If an absolute value is inferior to this threshold, it will equals zero. 
+ * @param {number} [options.thresholdError] If the error is inferior to that threshold, we say that the linear combinaison found is accepted, so the row is dependent from other rows.
+ * @param {Matrix} the matrix which represents the dependencies between rows.
```
Comment (maasencioh) [inline src/linearDependencies.js]: I think that this is the `@return`

File: src/linearDependencies.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        thresholdValue = 10e-10,
+        thresholdError = 10e-10
+    } = options;
+
+    var results = [];
+    var n = A.rows;
+
+    for (var i = 0; i < n; i++) {
+        var b = Matrix.columnVector(A.getRow(i));
+        var Abis = A.subMatrixRow(xrange(n, i)).transpose();
```
Comment (maasencioh) [inline src/linearDependencies.js]: please use transposeView

**Comment by maasencioh** [comment]
ping @jajoe

**Comment by maasencioh** [comment]
@targos should I fix the conflict first on this branch or could I do it on the merge?

**Comment by targos** [comment]
locally: delete package-lock.json and run `npm install`

**Pull request merged**